### PR TITLE
Generate unidirectional fields in generated schema

### PR DESCRIPTION
**Pull request by craicoverflow**
## Motivation

https://github.com/aerogear/graphback/issues/614

## What

This change uses a one-to-many field in the data model to generate the related field in the opposing type and inputs.

Add generated fields to:

- types
- filters
- inputs 

## Why

This enables the developer to perform mutations and queries from both types with minimal configuration.

## Verification Steps

Set the following data model:

```gql
type User {
  id: ID!
  name: String!
  """
  @db.oneToMany: 'inUser'
  """
  inTransactions: [Transaction!]!
  """
  @db.oneToMany: 'outUser'
  """
  outTransactions: [Transaction!]!
}

type Transaction {
  id: ID!
  amount: Int!
  outUser: User
}
```

Run `graphback generate` and `graphback db` to generate your code and migrate your database structure.

Look at your generated schema in `generated.ts`. The `Transaction` type should have automatically added the `inUser` field.

```gql
type Transaction {
  id: ID!
  amount: Int!
  inUser: User!
  outUser: User
}
```

Run the following mutations and queries in order to verify functionality.

```gql
mutation createUserA {
  createUser(input:{name:"User A"}) {
    id
    name
  }
}

mutation createUserB {
  createUser(input:{name:"User B"}) {
    id
    name
  }
}

mutation createTransaction {
  createTransaction(input: {amount:100, inUserId:1, outUserId:2}) {
    id
    amount
    inUser {
      id
      name
    }
    outUser {
      id
      name
    }
  }
}

query findTransactions {
  findAllTransactions {
    id
    amount
    inUser {
      id
      name
    }
    outUser {
      id
      name
    }
  }
}

query getUsersAndTransactions {
  findAllUsers {
    id
    name
    inTransactions {
      id
      amount
    }
    outTransactions {
      id
      amount
    }
  }
}
```

### CODE REVIEW

**Comment by b1zzu** [comment]
:eyes:

File: packages/graphback-codegen-schema/src/targetSchemaContext.ts [hunk 1]
```diff
@@ -186,22 +215,22 @@ export const buildTargetContext = (input: InputModelTypeContext[]) => {
     return {
       "name": t.name,
       "type": 'input',
-      "fields": [...t.fields.filter((f: InputModelFieldContext) => f.type !== 'ID' && (!(f.isType && f.isArray) || !f.isType))
+      "fields": [...new Set([...t.fields.filter((f: InputModelFieldContext) => f.type !== 'ID' && (!(f.isType && f.isArray) || !f.isType))
```
Comment (wtrocki) [inline packages/graphback-codegen-schema/src/targetSchemaContext.ts]: While this code works nicely this slowly becomes really hard to maintain. 
My main focus will be to get the cleaner path for relationships. I will share with you the idea once I will come up with solid plan

**Comment by craicoverflow** [comment]
> Snapshot needs to be pushed (as it was deleted)

It looks like it was deleted, but it was not:

https://github.com/aerogear/graphback/blob/unidirectional-schema-generation/packages/graphback-codegen-schema/tests/snapshots/GraphQLSchemaCreatorTest.ts.snap

**Comment by b1zzu** [comment]
@craicoverflow I'm trying this future and I found a couple of issues

After I've tried to create a Transaction like this
```
mutation {
  createTransaction(input: { amount: 10, inUserId: 1, outUserId: 2 }) {
    id
  }
}
```
and the I query it
```
query {
  findAllTransactions {
    id
    amount
    inUser {
      name
    }
    outUser {
      name
    }
  }
}
```
I recieve this error:
```
{
  "errors": [
    {
      "message": "Cannot return null for non-nullable field Transaction.inUser.",
      "locations": [
        {
          "line": 5,
          "column": 5
        }
      ],
      "path": [
        "findAllTransactions",
        0,
        "inUser"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "Error: Cannot return null for non-nullable field Transaction.inUser.",
            "    at completeValue (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:560:13)",
            "    at completeValueCatchingError (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:495:19)",
            "    at resolveField (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:435:10)",
            "    at executeFields (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:275:18)",
            "    at collectAndExecuteSubfields (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:713:10)",
            "    at completeObjectValue (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:703:10)",
            "    at completeValue (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:591:12)",
            "    at completeValue (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:557:21)",
            "    at completeValueCatchingError (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:495:19)",
            "    at /home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:618:25",
            "    at Array.forEach (<anonymous>)",
            "    at forEach (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/iterall/index.js:83:25)",
            "    at completeListValue (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:614:24)",
            "    at completeValue (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:573:12)",
            "    at completeValue (/home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:557:21)",
            "    at /home/dbizzarr/Projects/github.com/aerogear/graphback/examples/generator-fullstack/server/node_modules/graphql/execution/execute.js:492:16"
          ]
        }
      }
    }
  ],
  "data": null
}
```
The transaction is created correctlly in the db but it fails to query it
image


Also, the client mutations and fragment don't have the inUser
```
export const TransactionFragment = gql`
  fragment TransactionFields on Transaction {
    id
    amount
    outUserId
  }
`
```
```
export const createTransaction = gql`
  mutation createTransaction($amount: Int!, $outUserId: Int) {
    createTransaction(input: {amount: $amount, outUserId: $outUserId}) {
      ...TransactionFields
    }
  }


  ${TransactionFragment}
`
```

**Comment by b1zzu** [comment]
@craicoverflow my fault, I've copied the model you pasted which was missing the `inUser`

this is the correct one:
```
type Transaction {
  id: ID!
  amount: Int!
  inUser: User!
  outUser: User!
}
```

**Comment by b1zzu** [comment]
It works perfectly :+1:

**Comment by craicoverflow** [comment]
> @craicoverflow my fault, I've copied the model you pasted which was missing the `inUser`
> 
> this is the correct one:
> 
> ```
> type Transaction {
>   id: ID!
>   amount: Int!
>   inUser: User!
>   outUser: User!
> }
> ```

It's actually not working correctly, and you've spotted this!

The idea was that you wouldn't need to add `Transaction.inUser` as this would be generated by the `@db.oneToMany` annotation on `User`.

**Comment by b1zzu** [comment]
then it's not working :rofl:

**Pull request merged**